src/smartactions/SchedulingBase.sol

```diff
@@ -0,0 +1,194 @@
+import {IERC20} from "forge-std/interfaces/IERC20.sol";
+import {IERC7579Execution} from "modulekit/Accounts.sol";
+import {ERC7579ExecutorBase} from "modulekit/Modules.sol";
+import "modulekit/core/sessionKey/ISessionValidationModule.sol";
+
+abstract contract SchedulingBase is
+    ERC7579ExecutorBase,
+    ISessionValidationModule
+{
+    error InvalidExecution();
+    error InvalidMethod(bytes4);
+    error InvalidValue();
+    error InvalidAmount();
+    error InvalidTarget();
+    error InvalidRecipient();
+
+    error InvalidJob();
+
+    event ExecutionAdded(address indexed smartAccount, uint128 indexed jobId);
+
+    event ExecutionTriggered(
+        address indexed smartAccount,
+        uint128 indexed jobId
+    );
+
+    event ExecutionCancelled(
+        address indexed smartAccount,
+        uint128 indexed jobId
+    );
+
+    mapping(address smartAccount => mapping(uint128 jobId => ExecutionConfig))
+        internal _executionLog;
+
+    mapping(address smartAccount => uint128 jobCount) internal _accountJobCount;
+
+    struct ExecutionConfig {
+        uint48 executeInterval;
+        uint16 numberOfExecutions;
+        uint16 numberOfExecutionsCompleted;
+        uint48 startDate;
+        uint160 sqrtPriceLimitX96;
+        bool isEnabled;
+        uint48 lastExecutionTime;
+        bytes executionData;
+    }
+
+    struct ExecutorAccess {
+        address sessionKeySigner;
+        uint128 jobId;
+    }
+
+    struct Params {
+        uint128 jobId;
+    }
+
+    function _isExecutionValid(uint128 jobId) internal view {
+        ExecutionConfig storage executionConfig = _executionLog[msg.sender][
+            jobId
+        ];
+
+        if (!executionConfig.isEnabled) {
+            revert InvalidExecution();
+        }
+
+        if (
+            executionConfig.lastExecutionTime +
+                executionConfig.executeInterval <
+            block.timestamp
+        ) {
+            revert InvalidExecution();
+        }
+        if (
+            executionConfig.numberOfExecutionsCompleted >=
+            executionConfig.numberOfExecutions
+        ) {
+            revert InvalidExecution();
+        }
+        if (executionConfig.startDate > block.timestamp) {
+            revert InvalidExecution();
+        }
+    }
+
+    modifier canExecute(uint128 jobId) {
+        _isExecutionValid(jobId);
+        _;
+    }
+
+    // abstract methohd to be implemented by the inheriting contract
+    function executeOrder(uint128 jobId) external virtual;
+
+    function createExecution(ExecutionConfig calldata data) internal {
+        uint128 jobId = _accountJobCount[msg.sender] + 1;
+
+        _executionLog[msg.sender][jobId] = ExecutionConfig({
+            numberOfExecutionsCompleted: data.numberOfExecutionsCompleted,
+            executeInterval: data.executeInterval,
+            numberOfExecutions: data.numberOfExecutions,
+            startDate: data.startDate,
```

Comment:
what happens if due to a frontend mistake, startData == 0.

Reply:
@zeroknots probably we should validate startDate >= block.timestamp .. wdyt?

Reply:
@zeroknots probably we should validate startDate >= block.timestamp